### ISSUE TITLE
Support async methods on the IActivate interface

### ISSUE THREAD

**Issue by nigel-sampson** [Caliburn-Micro/Caliburn.Micro#506]
A lot of the initial discussion took place in #245, however my thinking has evolved a bit from then.

The initial plan is as follows.

1. Move the `IActivate` interface to async only with sync looking like
``` csharp
Task ActivateAsync(CancellationToken cancellationToken);
```
I'm not 100% on the "Async" suffix, if this was a new codebase that was async first I'd leave it off, however due to migration it make it easier to show where there are breaking changes.

2. Update the protected methods on `Screen` that are overridden to the following
``` csharp
Task OnInitializeAsync(CancellationToken cancellationToken);
Task OnActivateAsync(CancellationToken cancellationToken);
```
3. Update the `ScreenExtensions` such as `TryActivate` in similar ways.
4. Add some helper extension methods that allow activation without a cancellation token.
5. Update the Conductor classes in a similar method. This is probably the larger piece of work as it means a lot of methods become async such as `ChangeActiveItem` and `ActivateItem`.

Anybody have thoughts on this?

**Comment by godrose** [comment]
@nigel-sampson 
1. I think the "Async" suffix is great for alpha and is worth removing in the final release. You introduce the breaking signature changing change anyway (if we remove the helper methods out of the equation)
2. Same here.
3. Same here.
4. Have you considered option of providing the default value such as `CancellationToken cancellationToken = null`?
5. What is the source of asynchronousity in the final implementation? Default Scheduler?
Can it be substituted somehow for tests or any other purpose? 
Right now in all my projects I have one source of asynchronousity - in the Model layer. View Model layer always runs on the UI thread and no concurrency issues can happen there. I feel like when the View Model Layer is out of  control some concurrency issues may arise...

**Comment by nigel-sampson** [comment]
1. Not a big fan of adding and then removing it. 
4. Did consider that (although would be `= default(CancellationToken)` since it's a value type. But default parameter values can cause some silliness in libraries that overloads don't have.
5. Because your view models call into the model layer they need to be async for similar reasons. One of the largest issues is the way the `OnInitialize`, `OnActivate` lifecycle happens when the former is `async void`. 

My thought process is to expose the "async" the whole way up the stack, making decisions about "this should be async void to hide stuff" will usually be the wrong decision for someone. Hence `ChangeActiveItemAsync` etc.

The main point is that the view models won't do many things async, just better enable you to do them.

There could be some interesting things we can add where if a conductor is switching between items it can cancel the activation of the previous child view model.

**Comment by devhb** [comment]
Just a note with some problems I encountered with my try of a full async api.

In my try the view is not shown till the async OnInitialize was run completly so I needed to make a change, (see [moved notification about the ActiveItem before calling Activate on the screen so the view is displayed before executing any code](https://github.com/devhb/Caliburn.Micro/commit/2decc680890dadfbd739b5b1a2278758f1810af6)).

Can be that the behavior was selfmade but want to share my observation.

**Comment by markgould** [comment]
Not a huge fan of the async suffix but can see your reasons for it. As far as updating the conductor classes, etc, that would be great. Replacing the CanClose callbacks with async methods would also simplify things a lot.

Are you planning on pushing alphas as you hit certain milestones? I have a new project that I'll be using CM 4 on and would like to help test and give feedback as you progress.

**Comment by nigel-sampson** [comment]
Thanks @devhb .

@markgould I'm not a huge fan of it either. If this was a green field project I'd do away with, but having a clear breaking change in terms of method name makes things a bit easier for long term support.

In terms of milestones, I'll probably release an alpha.2 when `IActivate` and `IDeactivate` have been migrated.

**Comment by devhb** [comment]
@nigel-sampson I am interested in alpha prereleases too so we can test it and give early feedback.

**Comment by godrose** [comment]
@nigel-sampson Hi. Any update on the 4.0.0 release/pre-release?

**Comment by nigel-sampson** [comment]
So I've pushed all the changes for `IActivate` become async first,  There's a few hacks in there (hopefully marked) until other interfaces have been migrated e.g. `IWindowManager` and `INavigationService`.

Would appreciate some eyes on it and what's missing or broken. I think there's some missing overloads (without a `CancellationToken`).

**Comment by markgould** [comment]
I'll start looking at it soon - are you going to push a new alpha to nuget or should we just work off of develop?

**Comment by nigel-sampson** [comment]
Given the hacks I don't think it's a good candidate for an alpha.  I suspect the next few will have to be done at the same time (`IDeactivate`, `IGuardClose` and `ICloseStrategy`) and then will re-evaluate whether there should be an alpha.

**Comment by neilt6** [comment]
@nigel-sampson Has there been any progress on another 4.0.0 alpha with asynchronous `IActivate` and `IDeactivate`? I'm trying to update some USB code to use UWP APIs, but they're async-only and using `async void` for `OnActivate()` and `OnDeactivate()` is really ruining my day.

**Comment by nigel-sampson** [comment]
Apologies, it's been a rough few months and work hasn't gone as quickly as I'd hoped. 

Looking to have to have the commits in place for `IDeactivate`, `IGuardClose`, `ICloseStrategy` and `IClose` in place by the end of this week.

I don't believe that would make it production ready (the things that call into these methods need to be promoted too), but would definitely be soliciting feedback at that point

**Comment by neilt6** [comment]
@nigel-sampson Not a problem, I'm glad to hear development is still underway and look forward to trying out the changes.

I'm using a custom `NavigatingConductor<T>` which implements `INavigationService`, so I should be able to work around any oddities.

**Comment by nigel-sampson** [comment]
Most of the interfaces on the `dev/4.0.0` branch are now converted across to be async and it's close to where I think we can do the next alpha.  

The bits left that can come in the next alpha are
- Window Manager
- Navigation Services
- Bootstrappers / Applications

Before we actually do the alpha I'd love for interested parties to check out the branch and have a play. One of the problems that contributed to this taking longer than hoped was getting confidence in consistent repeatable builds. Having others verify that they're having no build issues would be great.

Overall the journey to get there with the new project system was a bit of a journey but I'm happy to where we got to, however it does come with more breaking changes beyond just the async stuff.  These include

- Minimum version for Caliburn.Micro.Core upped to .NET standard 1.3
- Minimum version for Caliburn.Micro.Platform for WPF has been pushed to .NET Framework 4.6
- Minimum version for Caliburn.Micro.Platform for UWP has been pushed to 16299.
- We're now depending on Xamarin.Forms 3.0.0.561731.

Also assembly names are now related close to project name `Caliburn.Micro.Core` and `Caliburn.Micro.Platform`.

Having `Caliburn.Micro.Xamarin.Forms` as a separate package with it's own assembly and namespace I've come to the conclusion that this is failed experiment in over engineering to deal with having both Caliburn.Micro for Android while also supporting the Xamarin.Forms version at the same time.

By deprecating this assembly and collapsing back into Caliburn.Micro.Platform we can kill Platform.Core. Yay.

The samples have been updated to show some of this working.

Would love to get feedback.
